Modules/_dbmmodule.c

```diff
@@ -76,16 +84,22 @@ newdbmobject(const char *file, int flags, int mode)
 static void
 dbm_dealloc(dbmobject *dp)
 {
-    if ( dp->di_dbm )
+    if (dp->di_dbm) {
         dbm_close(dp->di_dbm);
-    PyObject_Del(dp);
+    }
+    PyTypeObject *tp = Py_TYPE(dp);
+    tp->tp_free(dp);
+    Py_DECREF(tp);
 }
 
 static Py_ssize_t
 dbm_length(dbmobject *dp)
 {
+    PyTypeObject *tp = Py_TYPE(dp);
+    _dbm_state *state =  PyType_GetModuleState(tp);
```

Comment:
I'm not sure that tp variable is worth it:

```suggestion
    _dbm_state *state = PyType_GetModuleState(Py_TYPE(dp));
```

nitpick: my proposed change also removed an useless whitespace :-D